[LLVMify F18] Replace the use std::ostream with LLVM streams llvm::ostream

CarolineConcatto opened the pull request:
[LLVMify F18] Replace the use std::ostream with LLVM streams llvm::ostream

This patch replaces the occurrence of std::ostream by llvm::raw_ostream,
because  LLVM Coding Standards[1] says that "All new code should use raw_ostream instead of ostream".[1]

As a consequence, this patch also replaces the use of:
   std:stringstream by llvm::raw_string_ostream
   std::ofstream by llvm::raw_fd_ostream
   std::cout by llvm::outs()   and
   std::cerr by llvm::errs()
where it is needed, because of the use of llvm::raw_ostream.

[1]https://llvm.org/docs/CodingStandards.html

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

DavidTruby commented on include/flang/Common/enum-set.h:
```diff
@@ -16,11 +16,11 @@
 
 #include "constexpr-bitset.h"
 #include "idioms.h"
+#include "llvm/Support/raw_ostream.h"
```
We can actually forward declare `llvm::raw_ostream` (unlike `std::ostream`) here.
So instead of including the header you can do:
```
namespace llvm {
class raw_ostream;
}
```
and then include the header in the cpp file instead for faster compile times.

This also applies for most of the headers below.

schweitzpgi replied:
+1

CarolineConcatto replied:
Ok.

CarolineConcatto replied:
This header has this method llvm::raw_ostream &Dump implemented in enum-set, so the compile complains if I remove the header. 
It is ok to have header + namespace?
llvm::raw_ostream &Dump(
      llvm::raw_ostream &o, std::string EnumToString(enumerationType)) const {
    char sep{'{'};
    IterateOverMembers([&](auto e) {
      o << sep << EnumToString(e);
      sep = ',';
    });

DavidTruby replied:
I think we should move this function definition to the corresponding cpp file instead and just leave the declaration here

CarolineConcatto replied:
Should we do this in This PR or should I make another one? There are more header files that were implemented this way? We need to merge this before merging with LLVM.

DavidTruby replied:
I think this should be quite a small change to this PR so you should just do it here in my opinion

CarolineConcatto replied:
We think that these changes may lead to further changes. For now, I will leave this task for later as we have a deadline for this PR.

DavidTruby commented on lib/Semantics/mod-file.h:
```diff
@@ -33,17 +33,24 @@ class ModFileWriter {
   void Write(const Symbol &);
   std::string GetAsString(const Symbol &);
   void PutSymbols(const Scope &);
   void PutSymbol(std::stringstream &, const Symbol &);
```
Can we remove this function?

DavidTruby commented on lib/Semantics/mod-file.h:
```diff
@@ -8,8 +8,8 @@
-
 #include "flang/Semantics/attr.h"
+#include "llvm/Support/raw_ostream.h"
 #include <sstream>
```
Can we remove this header?

CarolineConcatto replied:
This one is ok, as well as #include <sstream>.

DavidTruby commented on lib/Semantics/scope.cpp:
```diff
@@ -10,6 +10,7 @@
+#include "llvm/Support/raw_ostream.h"
 #include <algorithm>
 #include <memory>
 #include <sstream>
```
Can we remove this header?

CarolineConcatto replied:
I believe not because of this method:

DavidTruby commented on lib/Semantics/tools.cpp:
```diff
@@ -17,6 +17,7 @@
+#include "llvm/Support/raw_ostream.h"
 #include <algorithm>
 #include <set>
 #include <sstream>
```
Can we remove this header?

CarolineConcatto replied:
According to our talks it will need, because of this: 
 llvm::raw_string_ostream ss {buf};
ss << "node has not been analyzed:\n";

DavidTruby replied:
I meant the sstream header here not the llvm one

DavidTruby commented on runtime/descriptor.cpp:
```diff
@@ -145,11 +146,13 @@ void Descriptor::Check() const {
-    << reinterpret_cast<std::intptr_t>(raw_.base_addr) << std::dec << '\n';
+llvm::raw_ostream &Descriptor::Dump(llvm::raw_ostream &o) const {
+ o << "Descriptor @ 0x"; 
+ o.llvm::raw_ostream::write_hex(reinterpret_cast<std::intptr_t>(this)) 
```
Can this not just be o.write_hex? I don't know why the qualification would be necessary.
It also builds for me without the qualification.

DavidTruby commented on test/Decimal/quick-sanity-test.cpp:
```diff
@@ -14,11 +15,12 @@ union u {
   u.x = x;
-  return std::cout << "FAIL: 0x" << std::hex << u.u << std::dec;
+  llvm::outs() << "FAIL: 0x";
+ return llvm::outs().write_hex(u.u);
```
See here where it's used unqualified.

DavidTruby commented on test/Evaluate/intrinsics.cpp:
```diff
@@ -4,6 +4,7 @@
 #include "flang/Parser/provenance.h"
+#include "llvm/Support/raw_ostream.h"
 #include <initializer_list>
 #include <iostream>
```
Use of this header is forbidden in LLVM. We should be able to remove it now?

schweitzpgi replied:
This isn't library code. It's a test program to generate tests.

CarolineConcatto replied:
This file has this method:
 void DoCall(std::optional<DynamicType> resultType = std::nullopt, int rank = 0, bool isElemental = false)
That uses std::cout, as the compiler did not complain about it I did not change. For that reason also that this is there.

CarolineConcatto replied:
I can do another commit removing all "#include <iostream>" in this PR if you all think it is fine. 
ATM this PR removes all uses of std::ostream in the code. I can also remove all the std::cout.

CarolineConcatto replied:
Another point is that I needed to all llvm because of this method:
void DoCall that uses AsFortran method and all AsFortran method was changed to use llvm:raw_ostream

DavidTruby commented on test/Runtime/hello.cpp:
```diff
@@ -2,6 +2,7 @@
 #include "../../runtime/io-api.h"
+#include "llvm/Support/raw_ostream.h"
 #include <cstring>
 #include <iostream>
```
Ditto here.

schweitzpgi commented on lib/Evaluate/formatting.cpp:
```diff
@@ -14,13 +14,13 @@
 #include "flang/Parser/characters.h"
 #include "flang/Semantics/symbol.h"
-#include <sstream>
+#include "llvm/Support/raw_ostream.h"
```
Interesting that it included string stream.

CarolineConcatto replied:
I believe that the reason was because of this function:
std::string ExpressionBase<RESULT>::AsFortran() const {
  std::ostringstream ss;
I've removed with the new commit.

schweitzpgi commented on lib/Semantics/mod-file.h:
```diff
@@ -33,17 +33,24 @@ class ModFileWriter {
+  std::string uses_Buf;
+  std::string useExtraAttrs_Buf;
+  std::string decls_Buf;
+  std::string contains_Buf;
```
Do these buffers need to be part of this class? Or can they be provided by the calling context and destructed there?

CarolineConcatto replied:
As far as I understand yes, it needs to be here. But  I double-check with @DavidTruby and he agrees.

DavidTruby replied:
If we don't have these here, the strings won't persist across member function calls, and these functions rely on that behaviour sometimes. Otherwise we could just have the streams themselves also not be part of the class.

schweitzpgi replied:
To be clear, I wasn't suggesting the backing strings be constructed and destroyed in some interleaved way. Clearly they need to persist throughout the context for correct output. This seems like a fine place to put them as they'll have the same lifetime as the streams. On the other hand, the object that owns the streams had explicit ways to reset the buffers at one point, which suggested that a single writer object might be reused to generate multiple outputs.

schweitzpgi commented on runtime/descriptor.cpp:
```diff
@@ -172,9 +175,9 @@ std::size_t DescriptorAddendum::SizeInBytes() const {
-  o << "  derivedType @ 0x" << std::hex
-    << reinterpret_cast<std::intptr_t>(derivedType_) << std::dec << '\n';
+llvm::raw_ostream &DescriptorAddendum::Dump(llvm::raw_ostream &o) const {
+  o << "  derivedType @ 0x"; o.llvm::raw_ostream::write_hex(
```
You don't need the namespace and class name here.
Also maybe needs a clang-format pass.

schweitzpgi commented on test/Evaluate/real.cpp:
```diff
@@ -398,7 +400,9 @@ void subsetTests(int pass, Rounding rounding, std::uint32_t opds) {
-      std::stringstream ss, css;
+      std::string ssBuf, cssBuf;
+      llvm::raw_string_ostream ss {ssBuf};
+      llvm::raw_string_ostream css {cssBuf};
```
Make sure to run clang-format. (I believe it will squish out all whitespace around the braces.)

CarolineConcatto replied:
So, I did not removed iostream from this file and others because it did not interact with the other files that I needed to change. This file has this function:

CarolineConcatto replied:
Done

CarolineConcatto commented on include/flang/Evaluate/formatting.h:
```diff
@@ -9,44 +9,44 @@
 template<typename A, bool COPYABLE>
 auto operator<<(
-    std::ostream &o, const Fortran::common::Indirection<A, COPYABLE> &x)
+    llvm::raw_ostream &o, const Fortran::common::Indirection<A, COPYABLE> &x)
```
So this is another example where the header file implements a method.

CarolineConcatto replied:
This task can be visited later when we have a merged with LLVM

CarolineConcatto commented on include/flang/Parser/dump-parse-tree.h:
```diff
@@ -761,7 +760,8 @@ class ParseTreeDumper {
   template<typename T> std::string AsFortran(const T &x) {
-    std::ostringstream ss;
+    std::string buf;
+    llvm::raw_string_ostream ss {buf};
```
Another header file with a method implementation.

CarolineConcatto replied:
This task can be visited later when we have a merged with LLVM

CarolineConcatto commented on include/flang/Parser/parsing.h:
```diff
@@ -50,19 +50,19 @@ class Parsing {
   void ClearLog();
 
-  void EmitMessage(std::ostream &o, const char *at, const std::string &message,
+  void EmitMessage(llvm::raw_ostream &o, const char *at, const std::string &message,
```
Another header file with a header Implementation

CarolineConcatto replied:
This task can be visited later when we have a merged with LLVM

CarolineConcatto commented on runtime/descriptor.h:
```diff
@@ -20,13 +20,16 @@
 
 #include "derived-type.h"
 #include "type-code.h"
-#include "flang/ISO_Fortran_binding.h"
```
Adding it back.

CarolineConcatto replied:
Removed the wrong header

CarolineConcatto replied:
Done!

CarolineConcatto commented on test/Evaluate/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ add_executable(leading-zero-bit-count-test
 
 target_link_libraries(leading-zero-bit-count-test
   FortranEvaluateTesting
+  LLVMSupport
```
I've added LLVMSupport in all the target_link_libraries that used FortranEvaluateTesting, because FortranEvaluateTesting used testing.cpp that used llvm::outs() and llvm::errs().

kiranchandramohan commented on include/flang/Parser/dump-parse-tree.h:
```diff
@@ -761,7 +760,8 @@ class ParseTreeDumper {
   template<typename T> std::string AsFortran(const T &x) {
-    std::ostringstream ss;
+    std::string buf;
+    llvm::raw_string_ostream ss {buf};
```
Is there a formatting issue here?

DavidTruby replied:
What's wrong with this formatting?

kiranchandramohan replied:
Thre is a space between ss and the brace.

CarolineConcatto replied:
I did not run clang-format-7 on the header files. Only on the cpp.

kiranchandramohan commented on include/flang/Parser/source.h:
```diff
@@ -44,16 +47,16 @@ class SourceFile {
-  bool Open(std::string path, std::stringstream *error);
-  bool ReadStandardInput(std::stringstream *error);
+  bool Open(std::string path, llvm::raw_ostream *error);
+  bool ReadStandardInput(llvm::raw_ostream *error);
```
Why is this not raw_string_ostream?

DavidTruby replied:
There's no need for these functions to take the more specific type really, we may for example eventually want to call or test them with a different stream type.

They should be references rather than pointers though now that you point it out

kiranchandramohan replied:
OK. If that is the case then please keep it consistent or provide justification on why raw_string_stream is used in functions Open and ReadStandardInput in Parser/provenance.cpp and TokenSequence::Put in Parser/token-sequence.cpp

kiranchandramohan commented on include/flang/Parser/source.h:
```diff
@@ -44,16 +47,16 @@ class SourceFile {
 
 private:
-  bool ReadFile(std::string errorPath, std::stringstream *error);
+  bool ReadFile(std::string errorPath, llvm::raw_ostream *error);
```
raw_string_ostream?

kiranchandramohan commented on lib/Parser/source.cpp:
```diff
@@ -111,7 +112,7 @@ static std::size_t RemoveCarriageReturns(char *buffer, std::size_t bytes) {
 }
 
-bool SourceFile::Open(std::string path, std::stringstream *error) {
+bool SourceFile::Open(std::string path, llvm::raw_ostream *error) {
```
raw_string_ostream? below also.

kiranchandramohan commented on lib/Semantics/expression.cpp:
```diff
@@ -2502,7 +2503,7 @@ bool ExpressionAnalyzer::EnforceTypeConstraint(parser::CharBlock at,
   if (result) {
     if (auto type{result->GetType()}) {
-      if (type->category() != category) { // C885
+      if (type->category() != category) {  // C885
```
additional space before comment?

CarolineConcatto replied:
It was added by clang-format-7 for some reason.

CarolineConcatto replied:
I did not run clang-format in this file as it adds the extra space.

kiranchandramohan commented on lib/Semantics/mod-file.cpp:
```diff
@@ -186,7 +193,7 @@ void ModFileWriter::PutSymbols(const Scope &scope) {
 // procedures, type-bound generics, final procedures) which go to typeBindings.
 void ModFileWriter::PutSymbol(
-    std::stringstream &typeBindings, const Symbol &symbol) {
+    llvm::raw_ostream &typeBindings, const Symbol &symbol) {
```
raw_string_ostream?

CarolineConcatto replied:
We've decided to replace all std::stringstream  parameters  by llvm::raw_ostream ones.

schweitzpgi replied:
+1 for improved flexibility.

kiranchandramohan commented on lib/Semantics/mod-file.h:
```diff
@@ -33,17 +36,23 @@ class ModFileWriter {
   std::string GetAsString(const Symbol &);
   void PutSymbols(const Scope &);
-  void PutSymbol(std::stringstream &, const Symbol &);
+  void PutSymbol(llvm::raw_ostream &, const Symbol &);
```
raw_string_ostream?

kiranchandramohan commented on test/Evaluate/intrinsics.cpp:
```diff
@@ -88,17 +88,17 @@ struct TestCall {
+      llvm::outs() << '(';
     }
-    std::cout << ')' << std::endl;
+    llvm::outs() << ')' << '\n';
```
should there be a flush?

kiranchandramohan commented on test/Evaluate/ISO-Fortran-binding.cpp:
```diff
@@ -71,13 +69,12 @@ static void AddNoiseToCdesc(CFI_cdesc_t *dv, CFI_rank_t rank) {
+      << " attribute: " << static_cast<int>(attr)
+      << " type: " << static_cast<int>(ty) << " elem_len: " << eLen
+      << " rank: " << static_cast<int>(rank) << " extent: ";
+  llvm::outs().write_hex(reinterpret_cast<std::intptr_t>(eAddr)) << '\n';
```
should there be a flush?

CarolineConcatto replied:
Ok, lesson learn. If there was a std::endl we need to replace by '\n' and flush().

CarolineConcatto replied:
Done for this file and others as well.

tskeith commented on lib/Semantics/mod-file.h:
```diff
@@ -33,17 +36,23 @@ class ModFileWriter {
-  std::stringstream decls_;
-  std::stringstream contains_;
+  // Buffer to use with raw_string_ostream
+  std::string uses_Buf;
```
Should be `usesBuf_`.

CarolineConcatto replied:
Done

kiranchandramohan commented on lib/Semantics/mod-file.cpp:
```diff
@@ -157,13 +158,17 @@ std::string ModFileWriter::GetAsString(const Symbol &symbol) {
   all << '\n' << uses_.str();
-  uses_.str(""s);
+  uses_.flush();
+  uses_Buf.clear();
```
Can this be uses_.str().clear() ?

CarolineConcatto replied:
I really dont know, because when I grep on llvm  there is only one use of something.str().clear. But lots of something.clear() for llvm::raw_string_stream.
@DavidTruby what do you think?

CarolineConcatto replied:
Talking with @DavidTruby it looks either is fine. But maybe the one proposed by Kiran in cleaner.

CarolineConcatto replied:
Done!

kiranchandramohan commented on lib/Parser/source.cpp:
```diff
@@ -111,36 +112,36 @@ static std::size_t RemoveCarriageReturns(char *buffer, std::size_t bytes) {
   fileDescriptor_ = open(path.c_str(), O_RDONLY);
   if (fileDescriptor_ < 0) {
-    *error << "Could not open " << errorPath << ": " << std::strerror(errno);
+    error << "Could not open " << errorPath << ": " << std::strerror(errno);
```
std::strerror -> llvm::sys::StrError? 
A few other occurences as well.

CarolineConcatto replied:
So, the compiler did not complain about that, and I believe it belongs to #include <cerrno>. Moreover, there is nothing in the LLVM Code style talking about this function.
But maybe you are right because I do see llvm::sys::StrError being used in llvm-project.
@DavidTruby ??

CarolineConcatto replied:
Ok, so it looks there is nothing asking to change, but the llvm one is thread-safe. So, for this reason(being thread-safe)[1], I am replacing std::strerror by  llvm::sys::StrError.
[1]https://llvm.org/doxygen/Errno_8cpp_source.html

DavidTruby replied:
llvm::sys::StrError has the same visible behaviour, but is thread safe where possible. I think we should use it over std::strerror for that reason.

CarolineConcatto replied:
Done!

kiranchandramohan commented on lib/Parser/parsing.cpp:
```diff
@@ -94,13 +95,15 @@ void Parsing::DumpCookedChars(std::ostream &out) const {
 }
 
-void Parsing::Parse(std::ostream *out) {
+void Parsing::Parse(llvm::raw_ostream *out) {
```
Is it possible to remove the pointer here also?

CarolineConcatto replied:
The problem with this function is its initialization in parsing.h
void Parsing::Parse(llvm::raw_ostream *out = nulltptr)
And it is initialized like that because in lib/Parser/parsing.cpp:130 it calls Parse without any parameter:

DavidTruby replied:
Llvm has a raw_null_ostream that you can pass instead, and you can then remove the null pointer checking code and just always print to out as it will get discarded when printing go raw_null_ostream.

DavidTruby replied:
What I'm suggesting is that you could default the parameter to llvm::raw_null_ostream instead of having it as a pointer and defaulting it to null. The current solution with the pointer is probably fine though.

CarolineConcatto commented:
@kiranchandramohan, if it is fine I rather do the message update when I do the squash for this PR.

kiranchandramohan commented:
Yes, it is fine.

CarolineConcatto commented:
Ok it has failed but it is not a problem with the PR, but memory problem.  2 runs are ok and 2 failed

kiranchandramohan commented:
@CarolineConcatto Thanks for pointing out that using the strerror from llvm in the runtime will introduce a dependency on the LLVM Runtime. You are correct, please remove the llvm function from the runtime and remove LLVMSupport from runtime/CMakeLists.txt.

CarolineConcatto commented:
@sscalpone I believe I've done the squash now!

sscalpone commented:
@CarolineConcatto Conflicts must be resolved.

isuruf commented on runtime/descriptor.h:
```diff
@@ -28,6 +28,10 @@
+namespace llvm {
+class raw_ostream;
+}
+
```
Should these be removed?

CarolineConcatto replied:
Done!

tskeith commented on tools/f18/f18.cpp:
```diff
@@ -123,14 +125,14 @@ bool ParentProcess() {
-  std::cerr << "execvp(" << argv[0] << ") failed: " << std::strerror(errno)
-            << '\n';
+  llvm::errs() << "execvp(" << argv[0] << ") failed: " << llvm::sys::StrError(errno)
+               << '\n';
```
It looks like this needs reformatting. Maybe other files too.

CarolineConcatto replied:
This is clang-format-7! I can re-run to see if it will change.

CarolineConcatto replied:
Done!

sscalpone commented:
A quick grep shows a few instances of #include <ostream>, e.g. message.h.

sscalpone commented:
Does the comment referencing ostream and cerr need adjustment?

CarolineConcatto commented:
Working on this again! Probably they were introduced by Friday. It would be nice if we could merge this, otherwise, I will need to re-do the works every-time there is a merge.

CarolineConcatto commented:
@sscalpon I've changed the message a little bit. I don't know if that is what you want, but let me know. I squashed also the patch and we should not have any conflict by now!

CarolineConcatto commented:
Memory problems again make it fail some tests, but the rest are ok!

CarolineConcatto commented:
Hey @sscalpone,
Sorry for me asking but, what dump.h?
I could not find any file called dump.h I could find this one:
include/flang/Parser/dump-parse-tree.h.
If I can remember, I needed to change because of AsFortran method. It uses stringstream and the compiler was complaining. I could not find any use of std:cerr in this file too. 
So I believe I am looking to the wrong file. Do you mind to point me to what file are you seeing the problem?

sscalpone commented:
Oops.

./tools/f18/dump.cpp:// reference and pointer, and for dumping to a provided ostream or cerr.

From: CarolineConcatto <notifications@github.com>
Reply-To: flang-compiler/f18 <reply@reply.github.com>
Date: Tuesday, March 17, 2020 at 1:30 AM
To: flang-compiler/f18 <f18@noreply.github.com>
Cc: Steve Scalpone <sscalpone@nvidia.com>, Mention <mention@noreply.github.com>
Subject: Re: [flang-compiler/f18] [LLVMify F18] Replace the use std::ostream with LLVM streams llvm::ostream (#1047)


Hey @sscalpone<https://github.com/sscalpone>,
Sorry for me asking but, what dump.h?
I could not find any file called dump.h I could find this one:
include/flang/Parser/dump-parse-tree.h.
If I can remember, I needed to change because of AsFortran method. It uses stringstream and the compiler was complaining. I could not find any use of std:cerr in this file too.
So I believe I am looking to the wrong file. Do you mind to point me to what file are you seeing the problem?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/flang-compiler/f18/pull/1047#issuecomment-599942369>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAD3D4XYJJ37YIXSJVB36ETRH4YKLANCNFSM4LAHSLKA>.

-----------------------------------------------------------------------------------
This email message is for the sole use of the intended recipient(s) and may contain
confidential information.  Any unauthorized review, use, disclosure or distribution
is prohibited.  If you are not the intended recipient, please contact the sender by
reply email and destroy all copies of the original message.
-----------------------------------------------------------------------------------

CarolineConcatto commented:
Just in case:
There are std::stringstream or #include <sstream>  in flatten.cpp. This file is not the compiler, is Peter's squash program. 
std::cout will be in the documentation/ImplementingASemanticCheck.md. Which I believe should not be changed because they explain how someone had implemented the semantic checks in the compiler.
std::cerr will be in documentation/ParserCombinators.md. Also another documentation file, that maybe needs change.

CarolineConcatto commented:
Hey @sscalpone I can change that message np.

sscalpone commented on documentation/ParserCombinators.md:
```diff
@@ -160,5 +160,5 @@ is built.  All of the following parsers consume characters acquired from
 ### Debugging Parser
 Last, a string literal `"..."_debug` denotes a parser that emits the string to
-`std::cerr` and succeeds.  It is useful for tracing while debugging a parser but should
+`llv::errs` and succeeds.  It is useful for tracing while debugging a parser but should
```
Is this really llv?  Sorry I didn't look it up.

CarolineConcatto replied:
Sorry @sscalpone. It looks like I cannot differentiate 2 l's from one l anymore. :(
I've looked too much for this.

CarolineConcatto replied:
Done

sscalpone commented:
@CarolineConcatto Conflicts need resolving.

sscalpone commented:
Is the comment in `tools/f18/dump.cpp` that mentions ostream still accurate?  I don't know if you can pass ostream to raw_ostream.

CarolineConcatto commented:
I believe the comment is about void Dump method, right?
It is overloaded. That method can dump raw_ostream or llvm:errs.

I was writing that this method dumps still dumps raw_ostream, that I would still leave as ostream. But by the confusion I've made, it is better to replace ostream by raw_ostream to avoid further confusion.

sscalpone commented:
@CarolineConcatto There's another conflict :(

CarolineConcatto commented:
@sscalpone conflict solved!

CarolineConcatto commented:
Again 1 test configuration out of 4 failed. Patch is still good!
